core/src/main/scala/com/evolutiongaming/catshelper/ReadWriteRef.scala

```diff
@@ -71,124 +71,166 @@ object ReadWriteRef {
+     */
+    case class State(pending: Queue[Pending] = Queue.empty, v: Long = 0) {
+      self =>
+
+      def withoutRead(v: Long): State = {
+        updateIndex(v) {
+          case r: Pending.Read  => if (r.active > 1) Some(r.decActive) else None
+          case w: Pending.Write => Some(w.copy(nextRead = w.nextRead.decActive))
         }
-        .flatten
+      }
 
-      val read: Resource[F, A] = {
-        // `suspend` makes it possible to cancel `.use` that waits for resource availability.
-        Resource.suspend {
-          Concurrent.cancelableF[F, Resource[F, A]] { cb =>
-            // Listener emits a resource when it's ready for immediate use.
-            val listener = (a: A) => {
-              val r = Resource { F.pure(a -> release(a, _ - 1)) }
-              F.delay(cb(r.asRight))
-            }
+      def withoutWrite(v: Long): State = {
+        updateIndex(v) {
+          case w: Pending.Write => if (w.nextRead.active > 0) Some(w.nextRead) else None
+          case unchanged        => Some(unchanged)
+        }
+      }
 
-            stateRef
-              .modify {
-                case s0 if s0.canRead =>
-                  val s1 = s0.copy(inUse = s0.inUse + 1)
-                  val effect = listener(s0.a)
-                  val cancel = F.unit
-                  s1 -> (effect as cancel)
-
-                case s0 => // have to wait for writer
-                  val pending = PendingRead(listener)
-                  val s1 = s0.copy(pending = s0.pending :+ pending)
-                  val effect = F.unit
-                  val cancel = stateRef.update(_.without(pending))
-                  s1 -> (effect as cancel)
-              }
-              .flatten
+      private def updateIndex(v: Long)(f: Pending => Option[Pending]): State = {
+        // Versions grow monotonically, so we can stop as soon as we find one or go beyond it
+        // Note that we don't account for the overflow yet. But even with increments happening
+        // each nanosecond we have 292 years to hit it.
+        @tailrec def loop(head: Queue[Pending], q: Queue[Pending]): State = {
+          q match {
+            case (p: Pending) +: tail if p.v < v  => loop(head :+ p, tail)
+            case (p: Pending) +: tail if p.v == v => self.copy(pending = head ++: (f(p) ++: tail))
+            case _                                => self
           }
         }
+        loop(Queue.empty, self.pending)
       }
+    }
 
-      val write: Resource[F, Upd[F, A]] = {
-        Resource.suspend {
-          Concurrent.cancelableF[F, Resource[F, Upd[F, A]]] { cb =>
-            val listener = (a: A) => {
-              val r = Resource {
-                Ref[F].of(a) map { aRef =>
-                  val upd: Upd[F, A] = (f: A => F[A]) => aRef.get.flatMap(f).flatTap(aRef.set)
-                  val commit = aRef.get.flatMap(a => release(a, _ => 0))
-                  upd -> commit
+    (Ref[F].of(init), Ref[F].of(State())) mapN { (aRef, stateRef) =>
```

Comment:
could you please explain on why you need two separate Refs ?

Reply:
Separation of concerns mostly. I tried to keep the `stateRef` only for read/write coordination to make it easier to understand (at least for myself).

Probably we can fit `a: A` back into the `State` but I'm not sure I want to do this right now. Do you see any problems with a separate ref?